ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,396 @@
+            open_content: self.open_content,
+            phantom: Default::default(),
+        }
+    }
+
+    pub fn id(&self) -> String {
+        self.id.to_owned()
+    }
+
+    pub fn imports(&self) -> &[IslImport] {
+        &self.imports
+    }
+
+    pub fn types(&self) -> &[IslType<V>] {
+        &self.types
+    }
+
+    pub fn inline_imported_types(&self) -> &[IslImportType] {
+        &self.inline_imported_types
+    }
+
+    /// Provides top level open content for given schema
+    /// For open content defined within type definitions use IslType#open_content()
+    pub fn open_content(&self) -> &Vec<Element> {
+        &self.open_content
+    }
+
+    /// Provide user reserved field defined in the given schema for ISL 2.0,
+    /// Otherwise returns None
+    pub fn user_reserved_fields(&self) -> Option<&UserReservedFields<V>> {
+        self.user_reserved_fields.as_ref()
+    }
+}
+
+/// Represents both named and anonymous [IslType]s and can be converted to a resolved type definition
+/// Named ISL type grammar: `type:: { name: <NAME>, <CONSTRAINT>...}`
+/// Anonymous ISL type grammar: `{ <CONSTRAINT>... }`
+#[derive(Debug, Clone, PartialEq)]
+pub struct IslType<V: IslVersionTrait> {
+    name: Option<String>,
+    constraints: Vec<IslConstraint<V>>,
+    // Represents the ISL type struct in string format for anonymous type definition
+    // For named type definition & programmatically created type definition, this will be `None`
+    pub(crate) isl_type_struct: Option<Element>,
+    phantom: PhantomData<V>,
+}
+
+impl<V: IslVersionTrait> IslType<V> {
+    pub(crate) fn new(name: Option<String>, constraints: Vec<IslConstraint<V>>) -> Self {
+        Self {
+            name,
+            constraints,
+            isl_type_struct: None,
+            phantom: Default::default(),
+        }
+    }
+
+    /// Creates a named [IslType] using the [IslConstraint] defined within it
+    pub fn named<A: Into<String>, B: Into<Vec<IslConstraint<V>>>>(
+        name: A,
+        constraints: B,
+    ) -> IslType<V> {
+        IslType::new(Some(name.into()), constraints.into())
+    }
+
+    /// Creates an anonymous [IslType] using the [IslConstraint] defined within it
+    pub fn anonymous<A: Into<Vec<IslConstraint<V>>>>(constraints: A) -> IslType<V> {
+        IslType::new(None, constraints.into())
+    }
+
+    /// Provides a name if the ISL type is named type definition
+    /// Otherwise returns None
+    pub fn name(&self) -> &Option<String> {
+        &self.name
+    }
+
+    /// Provides open content that is there in the type definition
+    pub fn open_content(&self) -> Vec<(&String, &Element)> {
+        let mut open_contents = vec![];
+        for constraint in &self.constraints {
+            if let IslConstraint::Unknown(open_content) = constraint {
+                open_contents.push((open_content.field_name(), open_content.field_value()))
+            }
+        }
+        open_contents
+    }
```

Comment:
I think this is a hint that open content should not be modeled as a constraint. Perhaps something like this instead?

```rust
pub struct IslType<V: IslVersionTrait> {
    name: Option<String>,
    constraints: Vec<IslConstraint<V>>,
    open_content: Vec<OpenContent<V>>,
    pub(crate) isl_type_struct: Option<Element>,
    phantom: PhantomData<V>,
}
```

Reply:
+1, I encountered some mental dissonance when I found `OpenContent` in the `IslConstraint` enum earlier in my review. Separating it would be helpful.